File: Web/Plugins/Lemoine.Plugin.ProductionTracker/CacheDomainExtension.cs
// Copyright (C) 2009-2023 Lemoine Automation Technologies
//
// SPDX-License-Identifier: Apache-2.0

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Threading.Tasks;
using Lemoine.Core.Cache;
using Lemoine.Core.Log;
using Lemoine.Extensions.Business.Cache;
using Lemoine.ModelDAO;

namespace Lemoine.Plugin.ProductionTracker
{
  public class CacheDomainExtension
    : Lemoine.Extensions.NotConfigurableExtension
    , ICacheDomainExtension
  {
    readonly ILog log = LogManager.GetLogger (typeof (CacheDomainExtension).FullName);

    public bool ClearDomain (ICacheClient cacheClient, string domain)
    {
      return false;
    }

    public bool ClearDomainByMachine (ICacheClient cacheClient, string domain, int machineId)
    {
      var regexes = GetCacheKeyRegexByMachine (domain, machineId);
      return cacheClient.ClearDomainFromRegexes (regexes);
    }

    public bool ClearDomainByMachineModule (ICacheClient cacheClient, string domain, int machineModuleId)
    {
      return false;
    }

    public IEnumerable<string> GetCacheKeyRegexByMachine (string domain, int machineId)
    {
      switch (domain.ToLowerInvariant ()) {
      case "processmachinestatetemplate":
      case "machineobservationstateassociation":
      case "business.operation.partproductionrange":
        return new List<string> {
          $@"ProductionTracker([?](.*&)?|/.*[&?])GroupId={machineId}(&.*)|$",
          $@"ProductionTracker([?](.*&)?|/.*[&?])GroupId=\d*[a-zA-Z].*(&.*)|$", // All the group ids that are not numeric
        };
      default:
        if (log.IsDebugEnabled) {
          log.Debug ($"GetCacheKeyRegexByMachine: unknown domain {domain}");
        }
        return new List<string> { };
      }
    }

    public async Task<bool> ClearDomainAsync (ICacheClient cacheClient, string domain)
    {
      return await Task .Run (() => ClearDomain (cacheClient, domain));
    }

    public async Task<bool> ClearDomainByMachineAsync (ICacheClient cacheClient, string domain, int machineId)
    {
      return await Task.Run (() => ClearDomainByMachine (cacheClient, domain, machineId));
    }

    public async Task<bool> ClearDomainByMachineModuleAsync (ICacheClient cacheClient, string domain, int machineModuleId)
    {
      return await Task.Run (() => ClearDomainByMachineModule (cacheClient, domain, machineModuleId));
     }
  }
}

File: Web/Plugins/Lemoine.Plugin.MachiningProcessMilestones/CacheDomainExtension.fs
(*
 * Copyright (C) 2009-2023 Lemoine Automation Technologies
 *
 * SPDX-License-Identifier: Apache-2.0
 *)

namespace Lemoine.Plugin.MachiningProcessMilestones

open System.Threading.Tasks

open Lemoine.Core.Log
open Lemoine.Core.Cache
open Lemoine.Extensions.Business.Cache

type CacheDomainExtension() =
  let log = LogManager.GetLogger ("Lemoine.Plugin.MachiningProcessMilestones.CacheDomainExtension")

  member val UniqueInstance = true

  member this.ClearDomain (cacheClient: ICacheClient, domain: string) =
    match domain.ToLowerInvariant () with
    | "plugin.machiningprocessmilestones" -> cacheClient.ClearDomainFromRegexes [| "Milestones.*" |]
    | _ -> false

  interface Lemoine.Extensions.IExtension with
    member this.UniqueInstance = this.UniqueInstance

  interface Lemoine.Extensions.Business.Cache.ICacheDomainExtension with
    member this.ClearDomain (cacheClient, domain) = this.ClearDomain (cacheClient, domain)
    member this.ClearDomainByMachine (cacheClient, domain, machineId) = false
    member this.ClearDomainByMachineModule (cacheClient, domain, machineModuleId) = false
    member this.ClearDomainAsync (cacheClient, domain) =
      task {
        return this.ClearDomain (cacheClient, domain)
      }      
    member this.ClearDomainByMachineAsync (cacheClient, domain, machineId) =
      task {
        return false
      }      
    member this.ClearDomainByMachineModuleAsync (cacheClient, domain, machineModuleId) =
      task {
        return false
      }
